### ISSUE TITLE
Custom `dbg!` macros for `dbg_macro` lint

### ISSUE THREAD

**Issue by ojeda** [rust-lang/rust-clippy#11303]
Back in Rust 1.60.0 we added `-Wclippy::dbg_macro` to the Linux kernel compilation flags, which worked great with our custom `dbg!` macro (essentially the `std` one but calling the kernel printing facilities).

However, in the next version, 1.61.0, as well as 1.71.1 and the current nightly, does not work. Thus only the first of the following lines emit a warning:

```sh
echo 'macro_rules! dbg { () => {} } fn main() { dbg!(); }' | rustup run  1.60.0 clippy-driver -Wclippy::dbg_macro -
echo 'macro_rules! dbg { () => {} } fn main() { dbg!(); }' | rustup run  1.61.0 clippy-driver -Wclippy::dbg_macro -
echo 'macro_rules! dbg { () => {} } fn main() { dbg!(); }' | rustup run  1.71.1 clippy-driver -Wclippy::dbg_macro -
echo 'macro_rules! dbg { () => {} } fn main() { dbg!(); }' | rustup run nightly clippy-driver -Wclippy::dbg_macro -
```

Is the lint intended to work only with the standard library `dbg` macro? I imagine that is the case, given the move to diagnostic items in https://github.com/rust-lang/rust-clippy/pull/7466, but it is not entirely clear from the lint description.

A workaround is to use `disallowed_macros`, but that does not have a specialized diagnostic message, does not have the `help:` suggestion, etc. Another is to use `rustc_attrs`, but I imagine that is not going to be stable.

Thus, instead, could the `dbg_macro` lint take a configuration with paths to `dbg!` macros (like `disallowed_macros`) or, even better, could there be an attribute that users could apply to their `dbg!` macro (like `rustc_diagnostic_item`)?

Cc @xFrednet and @Alexendoo who both seemed to work on this in the past.

**Comment by ojeda** [comment]
If one gets that level of customization, then it would be fine. Though if we wanted it to be as flexible, we would also need to be able to customize the lint level per path (which may be desirable anyway), since right now a project may put a different level to `dbg_macro` vs. `disallowed_macros`.
